lib/components/joystick/joystick_directional.dart

```diff
@@ -0,0 +1,290 @@
+    this.margin = const EdgeInsets.only(left: 100, bottom: 100),
+    this.size = 80,
+    this.color = Colors.blueGrey,
+    this.opacityBackground = 0.5,
+    this.opacityKnob = 0.8,
+  }) {
+    if (spriteBackgroundDirectional != null) {
+      _backgroundSprite = spriteBackgroundDirectional;
+    } else {
+      _paintBackground = Paint()
+        ..color = color.withOpacity(opacityBackground)
+        ..style = PaintingStyle.fill;
+    }
+    if (spriteKnobDirectional != null) {
+      _knobSprite = spriteKnobDirectional;
+    } else {
+      _paintKnob = Paint()
+        ..color = color.withOpacity(opacityKnob)
+        ..style = PaintingStyle.fill;
+    }
+
+    _tileSize = size / 2;
+  }
+
+  void initialize(Size _screenSize, JoystickController joystickController) {
+    this._screenSize = _screenSize;
+    _joystickController = joystickController;
+    final Offset osBackground =
+        Offset(margin.left, _screenSize.height - margin.bottom);
+    _backgroundRect = Rect.fromCircle(center: osBackground, radius: size / 2);
+
+    Offset osKnob =
+        Offset(_backgroundRect.center.dx, _backgroundRect.center.dy);
+    _knobRect = Rect.fromCircle(center: osKnob, radius: size / 4);
+
+    _dragPosition = _knobRect.center;
+  }
+
+  void render(Canvas canvas) {
+    if (_backgroundRect != null) {
+      if (_backgroundSprite != null) {
+        _backgroundSprite.renderRect(canvas, _backgroundRect);
+      } else {
+        final double radiusBackground = _backgroundRect.width / 2;
+        canvas.drawCircle(
+          Offset(_backgroundRect.left + radiusBackground,
+              _backgroundRect.top + radiusBackground),
+          radiusBackground,
+          _paintBackground,
+        );
+      }
+    }
+
+    if (_knobRect != null) {
+      if (_knobSprite != null) {
+        _knobSprite.renderRect(canvas, _knobRect);
+      } else {
+        final double radiusKnob = _knobRect.width / 2;
+        canvas.drawCircle(
+          Offset(_knobRect.left + radiusKnob, _knobRect.top + radiusKnob),
+          radiusKnob,
+          _paintKnob,
+        );
+      }
+    }
+  }
+
+  void update(double t) {
+    if (_dragging) {
+      final double _radAngle = atan2(
+          _dragPosition.dy - _backgroundRect.center.dy,
+          _dragPosition.dx - _backgroundRect.center.dx);
+
+      final double degrees = _radAngle * 180 / pi;
+
+      // Distance between the center of joystick background & drag position
+      final Point centerPoint =
+          Point(_backgroundRect.center.dx, _backgroundRect.center.dy);
+      double dist =
+          centerPoint.distanceTo(Point(_dragPosition.dx, _dragPosition.dy));
+
+      // The maximum distance for the knob position the edge of
+      // the background + half of its own size. The knob can wander in the
+      // background image, but not outside.
+      dist = dist < (_tileSize * _backgroundAspectRatio / 3)
+          ? dist
+          : (_tileSize * _backgroundAspectRatio / 3);
```

Comment:
You can assign `_tileSize * _backgroundAspectRatio / 3)` to a variable and avoid making the multiplication and division twice per update.

Reply:
even better, you can use the min/max functions